### Ejemplo 28
Use el método del punto fijo para determinar la solución del siguiente sistema, note que una solución es $x=2$ y $y=3$. Inicie el calculo suponiendo que $x=1.5$ y $y=3.5$. 
$$
    \left\{
    \begin{aligned}
    x^2+xy=10\\
    y+3xy^2=57
    \end{aligned}
    \right.
$$

_Solución:_ Despejando de las dos ecuaciones dadas, tenemos:
$$x_{i+1}=\frac{10-x_i^2}{y_i} \quad y_{i+1}=57-3x_iy_i^2$$

In [1]:
def f(x,y):
    return (10-x**2)/y
def g(x,y):
    return 57-3*x* y**2
def fixPoint2 (it,f,g,x,y):
    for i in range(it):
        r=f(x,y)
        x=r
        s=g(x,y)
        y=s
        print("Iteración",i+1,":")
        print("x=",x)
        print("y=",y,"\n")

In [2]:
fixPoint2(3,f,g,1.5,3.5)

Iteración 1 :
x= 2.2142857142857144
y= -24.375 

Iteración 2 :
x= -0.20910518053375193
y= 429.7136479591836 

Iteración 3 :
x= 0.02316955738026651
y= -12778.04178126569 



Note que con el despeje realizado, las soluciones divergen respecto a lo esperado. Despejando las mismas ecuaciones pero de otra forma, tenemos:
$$x_{i+1}=\sqrt{10-x_iy_i}\quad y_{i+1}=\sqrt{\frac{57-y_i}{3x_i}}$$

In [3]:
import math as m
def f(x,y):
    return m.sqrt(10-x*y)
def g(x,y):
    return m.sqrt((57-y)/(3*x))
fixPoint2(4,f,g,1.5,3.5)

Iteración 1 :
x= 2.179449471770337
y= 2.860505988116444 

Iteración 2 :
x= 1.9405338789116078
y= 3.0495506732218938 

Iteración 3 :
x= 2.0204562858807944
y= 2.9834047467413445 

Iteración 4 :
x= 1.9930281297863939
y= 3.005704362568321 



Note que en este caso, el método si parece converger al resultado deseado.

In [4]:
import numpy as np
def newtonRaphson2(f,x,tol=1.0e-9):
    " soln = newtonRaphson2(f,x,tol=1.0e-9).\n Resuelve un sistema de ecuaciones no lineales por el método de Newton-Raphson method usando x como suposición inicial. \n Note que f y x son vectores"
    def jacobian(f,x):
        h = 1.0e-4
        n = len(x)
        jac = np.zeros((n,n))
        f0 = f(x)
        for i in range(n):
            temp = x[i]
            x[i] = temp + h
            f1 = f(x)
            x[i] = temp
            jac[:,i] = (f1 - f0)/h
        return jac,f0
    
    for i in range(30):
        jac,f0 = jacobian(f,x)
        if m.sqrt(np.dot(f0,f0)/len(x)) < tol: 
            return x
        dx = np.linalg.solve(jac,-f0)
        x = x + dx
        if m.sqrt(np.dot(dx,dx)) < tol*max(max(abs(x)),1.0):
            return x
    print("Muchas iteraciones.")

### Ejemplo 29
Encuentre la solución de 
$$
    \left\{\begin{aligned}
    \sin(x)+y^2+\ln(z)=7\\
    3x+2^y-z^3=-1\\
    x+y+z=5
    \end{aligned}\right.
$$
Usando $\mathtt{newtonRaphson2}$. Use como suposiciones iniciales el punto (1,1,1).

_Solución:_ Sea $x_1=x$, $x_2=y$ y $x_3=z$.

In [5]:
def f(x):
    f = np.zeros(len(x))
    f[0] = m.sin(x[0]) + x[1]**2 + m.log(x[2]) - 7.0
    f[1] = 3.0*x[0] + 2.0**x[1] - x[2]**3 + 1.0
    f[2] = x[0] + x[1] + x[2] - 5.0
    return f
x = np.array([1.0, 1.0, 1.0])
sol=newtonRaphson2(f,x)
for i in range(len(sol)):
    print("x",i+1,"=",sol[i])

x 1 = 0.5990537566405691
x 2 = 2.3959314023778266
x 3 = 2.0050148409816035


### Ejercicio 31
Use el método de Newthon-Raphson para determinar la solución del sistema mostrado en el ejemplo #28.

_Solución:_ Sea $x_1=x$ y $x_2=y$.

In [6]:
def f(x):
    f = np.zeros(len(x))
    f[0] = x[0]**2 + x[0]*x[1] -10
    f[1] = x[1] + 3*x[0]* x[1]**2 -57
    return f
x = np.array([1.5, 3.5])
sol=newtonRaphson2(f,x)
for i in range(len(sol)):
    print("x",i+1,"=",sol[i])

x 1 = 2.0000000000020823
x 2 = 2.999999999990089


### Ejercicio 32
Determine los puntos de intersección entre el círculo $x^2+y^2=3$ y la hipérbola $xy=1$ sabiendo que una aproximación es (0.5,1.5).

_Solución:_ Sea $x_1=x$ y $x_2=y$.

In [7]:
def f(x):
    f = np.zeros(len(x))
    f[0] = x[0]**2 + x[1]**2 -3
    f[1] = x[1] * x[0] - 1 
    return f
x = np.array([0.5, 1.5])
sol=newtonRaphson2(f,x)
for i in range(len(sol)):
    print("x",i+1,"=",sol[i])

x 1 = 0.6180339887498757
x 2 = 1.618033988749945


Como las figuras son simétricas se tiene que los cuatro puntos de intersección vienen dados por $\pm(0.61803, 1.61803)$ y $\pm(1.61803, 0.61803)$.